ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  private void validateQueryTopicSources(final Query query) {
+    final QuerySourceTopicExtractor extractor = new QuerySourceTopicExtractor();
+    extractor.process(query, null);
+    for (String kafkaTopic : extractor.sourceTopics) {
+      checkAccess(kafkaTopic, AclOperation.READ);
+    }
+  }
+
+  private void validateCreateAsSelect(final CreateAsSelect createAsSelect) {
+    /*
+     * Check topic access for CREATE STREAM/TABLE AS SELECT statements.
+     *
+     * Validates Write on the target topic if exists, and Read on the query sources topics.
+     *
+     * The Create access is validated by the TopicCreateInjector which will attempt to create
+     * the target topic using the same ServiceContext used for validation.
+     */
+
+    validateQueryTopicSources(createAsSelect.getQuery());
+
+    String kafkaTopic;
+
+    try {
+      kafkaTopic = getSourceTopicName(createAsSelect.getName().getSuffix());
+    } catch (final KsqlException e) {
+      kafkaTopic = createAsSelect.getName().getSuffix();
+    }
+
+    if (serviceContext.getTopicClient().isTopicExists(kafkaTopic)) {
+      checkAccess(kafkaTopic, AclOperation.WRITE);
+    }
+  }
+
+  private void validateInsertInto(final InsertInto insertInto) {
+    /*
+     * Check topic access for INSERT INTO statements.
+     *
+     * Validates Write on the target topic, and Read on the query sources topics.
+     */
+
+    validateQueryTopicSources(insertInto.getQuery());
+
+    final String kafkaTopic = getSourceTopicName(insertInto.getTarget().getSuffix());
+    checkAccess(kafkaTopic, AclOperation.WRITE);
+  }
+
+  private String getSourceTopicName(final String ksqlTopicName) {
+    final DataSource<?> dataSource = metaStore.getSource(ksqlTopicName);
+    if (dataSource == null) {
+      throw new KsqlException("Cannot validate for topic access from an unknown stream/table: "
+          + ksqlTopicName);
+    }
+
+    return dataSource.getKafkaTopicName();
+  }
+
+  /**
+   * Checks if the ServiceContext has access to the topic with the specified AclOperation.
+   */
+  private void checkAccess(final String topicName, final AclOperation operation) {
+    final Set<AclOperation> authorizedOperations = serviceContext.getTopicClient()
+        .describeTopic(topicName).authorizedOperations();
+
+    if (!authorizedOperations.contains(operation)) {
+      // This error message is similar to what Kafka throws when it cannot access the topic
+      // due to an authorization error. I used this message to keep a consistent message.
+      throw new KsqlException(String.format(
+              "Failed to %s Kafka topic: [%s]%n"
+                  + "Caused by: Not authorized to access topic",
+              StringUtils.capitalize(operation.toString().toLowerCase()), topicName)
+      );
+    }
+  }
+
+  /**
+   * Helper class that extracts all source topics from a Query node
+   */
+  private final class QuerySourceTopicExtractor extends DefaultTraversalVisitor<Node, Void> {
```

Comment:
maybe leverage `SourceTopicExtractor` in `TopicCreateInjector`? (e.g. make it public and move it to top level)

Reply:
Done. I created a `SourceTopicsExtractor`